tests/Unit/MerchantCenter/AccountServiceTest.php

```diff
@@ -591,6 +598,88 @@ public function test_switch_url() {
 		$this->assertEquals( self::TEST_ACCOUNT_DATA, $this->account->switch_url( self::TEST_ACCOUNT_ID ) );
 	}
 
+	public function test_setup_account_step_link_ads() {
+		$this->options->expects( $this->any() )
+			->method( 'get_ads_id' )
+			->willReturn( self::TEST_ADS_ID );
+
+		$this->options->expects( $this->any() )
+			->method( 'get_merchant_id' )
+			->willReturn( self::TEST_ACCOUNT_ID );
+
+		$this->state->expects( $this->once() )
+			->method( 'get' )
+			->willReturn(
+				[
+					'link_ads' => [ 'status' => MerchantAccountState::STEP_PENDING ],
+				]
+			);
+
+		$this->merchant->expects( $this->once() )
+			->method( 'link_ads_id' )
+			->with( self::TEST_ADS_ID );
+
+		$this->ads->expects( $this->once() )
+			->method( 'accept_merchant_link' )
+			->with( self::TEST_ACCOUNT_ID );
+
+		$this->assertEquals( self::TEST_ACCOUNT_DATA, $this->account->setup_account( self::TEST_ACCOUNT_ID ) );
+	}
+
+	public function test_setup_account_step_link_ads_without_ads() {
+		$mc_account_state = [
+			'link_ads' => [ 'status' => MerchantAccountState::STEP_PENDING ],
+		];
+
+		$this->options->expects( $this->any() )
+			->method( 'get_ads_id' )
+			->willReturn( 0 );
+
+		$this->options->expects( $this->any() )
+			->method( 'get_merchant_id' )
+			->willReturn( self::TEST_ACCOUNT_ID );
+
+		$this->state->expects( $this->once() )
+			->method( 'get' )
+			->willReturn( $mc_account_state );
+
+		$this->state->expects( $this->once() )
+			->method( 'update' )
+			->with( $mc_account_state );
+
+		$this->middleware->expects( $this->never() )
+			->method( 'link_ads_account' );
+
+		$this->assertEquals( [ 'id' => self::TEST_ACCOUNT_ID ], $this->account->setup_account( self::TEST_ACCOUNT_ID ) );
+	}
+
+	public function test_setup_account_step_link_ads_without_mc() {
+		$mc_account_state = [
+			'link_ads' => [ 'status' => MerchantAccountState::STEP_PENDING ],
+		];
+
+		$this->options->expects( $this->any() )
+			->method( 'get_ads_id' )
+			->willReturn( self::TEST_ADS_ID );
+
+		$this->options->expects( $this->any() )
+			->method( 'get_merchant_id' )
+			->willReturn( 0 );
+
+		$this->state->expects( $this->once() )
+			->method( 'get' )
+			->willReturn( $mc_account_state );
+
+		$this->state->expects( $this->once() )
+			->method( 'update' )
+			->with( $mc_account_state );
+
+		$this->middleware->expects( $this->never() )
+			->method( 'link_ads_account' );
+
+		$this->assertEquals( [ 'id' => 0 ], $this->account->setup_account( self::TEST_ACCOUNT_ID ) );
+	}
```

Comment:
If we simplify the code to return an exception then this test can also be simplified and made more logical:

```suggestion
	public function test_setup_account_step_link_ads_without_mc() {
		$this->options->expects( $this->any() )
			->method( 'get_ads_id' )
			->willReturn( self::TEST_ADS_ID );

		$this->options->expects( $this->any() )
			->method( 'get_merchant_id' )
			->willReturn( 0 );

		$this->state->expects( $this->once() )
			->method( 'get' )
			->willReturn(
				[
					'link_ads' => [ 'status' => MerchantAccountState::STEP_PENDING ],
				]
			);

		$this->expectException( Exception::class );
		$this->expectExceptionMessage( 'A Merchant Center account must be connected' );

		$this->account->setup_account( self::TEST_ACCOUNT_ID );
	}
```